From 2b3138e5204941ef92bc8a939c8917b28ee571b3 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Tue, 25 Mar 2025 22:27:48 +0100
Subject: [PATCH 1662/1695] drivers: stm32_risaf: alignment with upstream

Backport all changes applied on upstream driver.
This will ease next modifications and prepare as much as possible
further upstream operations.
Align also platform side with PAI changes.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ddb020802b1e0acc7660b02e2adf9bf830ccf1eb drivers: stm32_risaf: alignment with upstream]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: Iad10caf619b7a26a8bee1b30f4819ac122316f9a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/446825
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c  |   2 +-
 core/drivers/firewall/stm32_risaf.c | 586 ++++++++++++----------------
 core/include/drivers/stm32_risaf.h  |  27 +-
 3 files changed, 258 insertions(+), 357 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 6b58a2c3c..a3c9ca607 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -156,7 +156,7 @@ void boot_secondary_init_intc(void)
 void stm32_rif_access_violation_action(void)
 {
 #ifdef CFG_STM32_RISAF
-	stm32_risaf_dump_erroneous_data();
+	stm32_risaf_print_erroneous_data();
 	stm32_risaf_clear_illegal_access_flags();
 #endif
 #ifdef CFG_STM32_RISAB
diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index 520b29b33..4474ff65e 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
- * Copyright (c) 2021-2023, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2021-2024, STMicroelectronics
  */
 
 #include <assert.h>
@@ -105,55 +105,42 @@
 #define _RISAF_REG_CIDCFGR_WRENC_MASK	GENMASK_32(23, 16)
 #define _RISAF_REG_CIDCFGR_ALL_MASK	(_RISAF_REG_CIDCFGR_RDENC_MASK | \
 					 _RISAF_REG_CIDCFGR_WRENC_MASK)
-#define _RISAF_REG_READ_OK(reg, cid)	((reg) & \
-					 (BIT((cid)) << \
-					  _RISAF_REG_CIDCFGR_RDENC_SHIFT))
-#define _RISAF_REG_WRITE_OK(reg, cid)	((reg) & \
-					 (BIT((cid)) << \
-					  _RISAF_REG_CIDCFGR_WRENC_SHIFT))
+#define _RISAF_REG_READ_OK(reg, cid) \
+	((reg) & BIT((cid) + _RISAF_REG_CIDCFGR_RDENC_SHIFT))
+#define _RISAF_REG_WRITE_OK(reg, cid)	\
+	((reg) & BIT((cid) + _RISAF_REG_CIDCFGR_WRENC_SHIFT))
 
 #define _RISAF_GET_REGION_ID(cfg)	((cfg) & DT_RISAF_REG_ID_MASK)
-#if defined(CFG_STM32MP21)
-#define _RISAF_GET_REGION_CFG(cfg) \
-	(((((cfg) & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT) << \
-	  _RISAF_REG_CFGR_BREN_SHIFT) | \
-	 ((((cfg) & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT) << \
-	  _RISAF_REG_CFGR_SEC_SHIFT) | \
-	 ((((cfg) & DT_RISAF_ENC_MASK) >> DT_RISAF_ENC_SHIFT) << \
-	  _RISAF_REG_CFGR_ENC_SHIFT) | \
-	 ((((cfg) & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT) << \
-	  _RISAF_REG_CFGR_PRIVC_SHIFT))
-#else /* defined(CFG_STM32MP21) */
-#define _RISAF_GET_REGION_CFG(cfg) \
-	(((((cfg) & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT) << \
-	  _RISAF_REG_CFGR_BREN_SHIFT) | \
-	 ((((cfg) & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT) << \
-	  _RISAF_REG_CFGR_SEC_SHIFT) | \
-	 ((((cfg) & DT_RISAF_ENC_MASK) >> (DT_RISAF_ENC_SHIFT + 1)) << \
-	  _RISAF_REG_CFGR_ENC_SHIFT) | \
-	 ((((cfg) & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT) << \
-	  _RISAF_REG_CFGR_PRIVC_SHIFT))
-#endif /* defined(CFG_STM32MP21) */
-#define _RISAF_GET_REGION_CID_CFG(cfg) \
-	(((((cfg) & DT_RISAF_WRITE_MASK) >> DT_RISAF_WRITE_SHIFT) << \
-	  _RISAF_REG_CIDCFGR_WRENC_SHIFT) | \
-	 ((((cfg) & DT_RISAF_READ_MASK) >> DT_RISAF_READ_SHIFT) << \
-	  _RISAF_REG_CIDCFGR_RDENC_SHIFT))
-
-#ifdef CFG_CORE_RESERVED_SHM
-#define TZDRAM_RESERVED_SIZE		(TZDRAM_SIZE + TEE_SHMEM_SIZE)
-#else
-#define TZDRAM_RESERVED_SIZE		TZDRAM_SIZE
-#endif
 
 #define _RISAF_NB_CID_SUPPORTED		U(8)
 
+/**
+ * struct stm32_risaf_region - RISAF memory region
+ *
+ * @addr: Region base address.
+ * @len: Length of the memory region.
+ * @cfg: Region configuration.
+ */
 struct stm32_risaf_region {
-	uint32_t cfg;
-	uintptr_t addr;
+	paddr_t addr;
 	size_t len;
+	uint32_t cfg;
 };
 
+/**
+ * struct stm32_risaf_pdata - RISAF platform data
+ *
+ * @base: Base address of the RISAF instance.
+ * @clock: Clock of the RISAF.
+ * @regions: Number of memory regions, defined by the device tree configuration.
+ * @risaf_name: Name of the RISAF instance
+ * @nregions: Number of memory regions found in the device tree.
+ * @conf_lock: State whether the RISAF configuration is locked.
+ * @mem_base: Base address of the memory range covered by the RISAF instance.
+ * @mem_size: Size of the memory range covered by the RISAF instance.
+ * @enc_supported: If true, the RISAF instance supports encryption of the memory
+ * regions.
+ */
 struct stm32_risaf_pdata {
 	struct io_pa_va base;
 	struct clk *clock;
@@ -161,11 +148,20 @@ struct stm32_risaf_pdata {
 	char risaf_name[20];
 	unsigned int nregions;
 	unsigned int conf_lock;
-	uintptr_t mem_base;
+	paddr_t mem_base;
 	size_t mem_size;
 	bool enc_supported;
 };
 
+/**
+ * struct stm32_risaf_ddata - RISAF driver data
+ *
+ * @mask_regions: Number of address bits to match when determining access to a
+ * base region or subregion (WIDTH).
+ * @max_base_regions: Number of subdivision of the memory range (A.K.A memory
+ * regions) supported by the RISAF instance.
+ * @granularity: Length of the smallest possible region size.
+ */
 struct stm32_risaf_ddata {
 	uint32_t mask_regions;
 	uint32_t max_base_regions;
@@ -187,10 +183,10 @@ struct stm32_risaf_version {
 };
 
 /**
- * struct stm32_risaf_compat_data describes RISAF associated data
+ * struct stm32_risaf_compat_data  - Describes RISAF associated data
  * for compatible list.
  *
- * @supported_encryption:	identify risaf encryption capabilities.
+ * @supported_encryption:	identify RISAF encryption capabilities.
  */
 struct stm32_risaf_compat_data {
 	bool supported_encryption;
@@ -214,6 +210,37 @@ static vaddr_t risaf_base(struct stm32_risaf_instance *risaf)
 	return io_pa_or_va_secure(&risaf->pdata.base, 1);
 }
 
+static uint32_t stm32_risaf_get_region_config(uint32_t cfg)
+{
+#if defined(CFG_STM32MP21)
+	return SHIFT_U32((cfg & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT,
+			 _RISAF_REG_CFGR_BREN_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT,
+			 _RISAF_REG_CFGR_SEC_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_ENC_MASK) >> (DT_RISAF_ENC_SHIFT),
+			 _RISAF_REG_CFGR_ENC_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT,
+			 _RISAF_REG_CFGR_PRIVC_SHIFT);
+#else /* defined(CFG_STM32MP21) */
+	return SHIFT_U32((cfg & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT,
+			 _RISAF_REG_CFGR_BREN_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT,
+			 _RISAF_REG_CFGR_SEC_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_ENC_MASK) >> (DT_RISAF_ENC_SHIFT + 1),
+			 _RISAF_REG_CFGR_ENC_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT,
+			 _RISAF_REG_CFGR_PRIVC_SHIFT);
+#endif /* defined(CFG_STM32MP21) */
+}
+
+static uint32_t stm32_risaf_get_region_cid_config(uint32_t cfg)
+{
+	return SHIFT_U32((cfg & DT_RISAF_WRITE_MASK) >> DT_RISAF_WRITE_SHIFT,
+			 _RISAF_REG_CIDCFGR_WRENC_SHIFT) |
+	       SHIFT_U32((cfg & DT_RISAF_READ_MASK) >> DT_RISAF_READ_SHIFT,
+			 _RISAF_REG_CIDCFGR_RDENC_SHIFT);
+}
+
 void stm32_risaf_clear_illegal_access_flags(void)
 {
 	struct stm32_risaf_instance *risaf = NULL;
@@ -221,75 +248,62 @@ void stm32_risaf_clear_illegal_access_flags(void)
 	SLIST_FOREACH(risaf, &risaf_list, link) {
 		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
 
-		if (clk_enable(risaf->pdata.clock))
-			panic("Can't enable RISAF clock");
-
-		if (!io_read32(base + _RISAF_IASR)) {
-			clk_disable(risaf->pdata.clock);
+		if (!io_read32(base + _RISAF_IASR))
 			continue;
-		}
 
 		io_write32(base + _RISAF_IACR, _RISAF_IACR_CAEF |
 			   _RISAF_IACR_IAEF0 | _RISAF_IACR_IAEF1);
-
-		clk_disable(risaf->pdata.clock);
 	}
 }
 
-#ifdef CFG_TEE_CORE_DEBUG
-void stm32_risaf_dump_erroneous_data(void)
+void stm32_risaf_print_erroneous_data(void)
 {
 	struct stm32_risaf_instance *risaf = NULL;
 
+	if (!IS_ENABLED(CFG_TEE_CORE_DEBUG))
+		return;
+
 	SLIST_FOREACH(risaf, &risaf_list, link) {
 		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
 
-		if (clk_enable(risaf->pdata.clock))
-			panic("Can't enable RISAF clock");
-
 		/* Check if faulty address on this RISAF */
-		if (!io_read32(base + _RISAF_IASR)) {
-			clk_disable(risaf->pdata.clock);
+		if (!io_read32(base + _RISAF_IASR))
 			continue;
-		}
 
-		EMSG("\n\nDUMPING DATA FOR %s\n\n", risaf->pdata.risaf_name);
-		EMSG("=====================================================");
-		EMSG("Status register (IAESR0): %#"PRIx32,
+		IMSG("\n\nDUMPING DATA FOR %s\n\n", risaf->pdata.risaf_name);
+		IMSG("=====================================================");
+		IMSG("Status register (IAESR0): %#"PRIx32,
 		     io_read32(base + _RISAF_IAESR0));
 
 		/* Reserved if dual port feature not available */
 		if (io_read32(base + _RISAF_IAESR1))
-			EMSG("Status register Dual Port (IAESR1) %#"PRIx32,
+			IMSG("Status register Dual Port (IAESR1) %#"PRIx32,
 			     io_read32(base + _RISAF_IAESR1));
 
-		EMSG("-----------------------------------------------------");
+		IMSG("-----------------------------------------------------");
 		if (virt_to_phys((void *)base) == RISAF4_BASE) {
-			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			IMSG("Faulty address (IADDR0): %#"PRIxPA,
 			     risaf->pdata.mem_base +
-			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+			     io_read32(base + _RISAF_IADDR0));
 
 			/* Reserved if dual port feature not available */
 			if (io_read32(base + _RISAF_IADDR1))
-				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
+				IMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
 				     risaf->pdata.mem_base +
-				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+				     io_read32(base + _RISAF_IADDR1));
 		} else {
-			EMSG("Faulty address (IADDR0): %#"PRIxPA,
-			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+			IMSG("Faulty address (IADDR0): %#"PRIx32,
+			     io_read32(base + _RISAF_IADDR0));
 
 			/* Reserved if dual port feature not available */
 			if (io_read32(base + _RISAF_IADDR1))
-				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
-				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+				IMSG("Dual port faulty address (IADDR1): %#"PRIx32,
+				     io_read32(base + _RISAF_IADDR1));
 		}
 
-		EMSG("=====================================================\n");
-
-		clk_disable(risaf->pdata.clock);
+		IMSG("=====================================================\n");
 	};
 }
-#endif
 
 static __maybe_unused
 bool risaf_is_hw_encryption_enabled(struct stm32_risaf_instance *risaf)
@@ -302,29 +316,20 @@ static TEE_Result
 risaf_check_region_boundaries(struct stm32_risaf_instance *risaf,
 			      struct stm32_risaf_region *region)
 {
-	uintptr_t end_paddr = 0;
-
-	assert(region->len != 0U);
-
-	if (region->addr < risaf->pdata.mem_base) {
-		EMSG("RISAF %#"PRIxPTR": region start address lower than memory base",
-		     risaf->pdata.base.pa);
-		return TEE_ERROR_BAD_PARAMETERS;
-	}
-
-	/* Get physical end address */
-	end_paddr = risaf->pdata.mem_base + (risaf->pdata.mem_size - 1U);
-	if (region->addr > end_paddr ||
-	    ((region->addr - 1U + region->len) > end_paddr)) {
-		EMSG("RISAF %#"PRIxPTR": start/end address higher than physical end",
-		     risaf->pdata.base.pa);
+	if (!core_is_buffer_inside(region->addr, region->len,
+				   risaf->pdata.mem_base,
+				   risaf->pdata.mem_size)) {
+		EMSG("Region %#"PRIxPA"..%#"PRIxPA" outside RISAF area %#"PRIxPA"...%#"PRIxPA,
+		     region->addr, region->addr + region->len - 1,
+		     risaf->pdata.mem_base,
+		     risaf->pdata.mem_base + risaf->pdata.mem_size - 1);
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
 	if (!risaf->ddata->granularity ||
 	    (region->addr % risaf->ddata->granularity) ||
 	    (region->len % risaf->ddata->granularity)) {
-		EMSG("RISAF %#"PRIxPTR": start/end address granularity not respected",
+		EMSG("RISAF %#"PRIxPA": start/end address granularity not respected",
 		     risaf->pdata.base.pa);
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
@@ -347,7 +352,7 @@ risaf_check_overlap(struct stm32_risaf_instance *risaf __maybe_unused,
 					     region[index].len,
 					     region[i].addr,
 					     region[i].len)) {
-			EMSG("RISAF %#"PRIxPTR": Regions %u and %u overlap",
+			EMSG("RISAF %#"PRIxPA": Regions %u and %u overlap",
 			     risaf->pdata.base.pa, index, i);
 			return TEE_ERROR_GENERIC;
 		}
@@ -358,29 +363,15 @@ risaf_check_overlap(struct stm32_risaf_instance *risaf __maybe_unused,
 
 static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
 					 uint32_t region_id, uint32_t cfg,
-					 uint32_t cid_cfg, uintptr_t saddr,
-					 uintptr_t eaddr)
+					 uint32_t cid_cfg, paddr_t saddr,
+					 paddr_t eaddr)
 {
-	vaddr_t base = risaf_base(risaf);
 	uint32_t mask = risaf->ddata->mask_regions;
-
-	io_clrbits32(base + _RISAF_REG_CFGR(region_id), _RISAF_REG_CFGR_BREN);
-
-	io_clrsetbits32(base + _RISAF_REG_STARTR(region_id), mask,
-			(saddr - risaf->pdata.mem_base) & mask);
-	io_clrsetbits32(base + _RISAF_REG_ENDR(region_id), mask,
-			(eaddr - risaf->pdata.mem_base) & mask);
-	io_clrsetbits32(base + _RISAF_REG_CIDCFGR(region_id),
-			_RISAF_REG_CIDCFGR_ALL_MASK,
-			cid_cfg & _RISAF_REG_CIDCFGR_ALL_MASK);
-
-	io_clrsetbits32(base + _RISAF_REG_CFGR(region_id),
-			_RISAF_REG_CFGR_ALL_MASK,
-			cfg & _RISAF_REG_CFGR_ALL_MASK);
+	vaddr_t base = risaf_base(risaf);
 
 	if (cfg & _RISAF_REG_CFGR_ENC) {
 		if (!risaf->pdata.enc_supported) {
-			EMSG("RISAF %#"PRIxPTR": encryption feature error",
+			EMSG("RISAF %#"PRIxPA": encryption feature error",
 			     risaf->pdata.base.pa);
 			return TEE_ERROR_GENERIC;
 		}
@@ -397,15 +388,32 @@ static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
 		}
 
 		if ((cfg & _RISAF_REG_CFGR_SEC) != _RISAF_REG_CFGR_SEC) {
-			EMSG("RISAF %#"PRIxPTR": encryption on non secure area",
+			EMSG("RISAF %#"PRIxPA": encryption on non-secure area",
 			     risaf->pdata.base.pa);
 			return TEE_ERROR_GENERIC;
 		}
 	}
 
-	DMSG("RISAF %#"PRIxPTR": region %02d - start 0x%08x - end 0x%08x - cfg 0x%08x - cidcfg 0x%08x",
+	io_clrbits32(base + _RISAF_REG_CFGR(region_id), _RISAF_REG_CFGR_BREN);
+
+	io_clrsetbits32(base + _RISAF_REG_STARTR(region_id), mask,
+			(saddr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_ENDR(region_id), mask,
+			(eaddr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_CIDCFGR(region_id),
+			_RISAF_REG_CIDCFGR_ALL_MASK,
+			cid_cfg & _RISAF_REG_CIDCFGR_ALL_MASK);
+
+	io_clrsetbits32(base + _RISAF_REG_CFGR(region_id),
+			_RISAF_REG_CFGR_ALL_MASK,
+			cfg & _RISAF_REG_CFGR_ALL_MASK);
+
+	DMSG("RISAF %#"PRIxPA": region %02"PRIu32" - start %#"PRIxPA
+	     "- end %#"PRIxPA" - cfg %#08"PRIx32" - cidcfg %#08"PRIx32,
 	     risaf->pdata.base.pa, region_id,
+	     risaf->pdata.mem_base +
 	     io_read32(base + _RISAF_REG_STARTR(region_id)),
+	     risaf->pdata.mem_base +
 	     io_read32(base + _RISAF_REG_ENDR(region_id)),
 	     io_read32(base + _RISAF_REG_CFGR(region_id)),
 	     io_read32(base + _RISAF_REG_CIDCFGR(region_id)));
@@ -425,50 +433,35 @@ static void risaf_print_version(struct stm32_risaf_instance *risaf)
 		.size_id = io_read32(base + _RISAF_SIDR)
 	};
 
-	DMSG("RISAF %#"PRIxPTR" version %d.%d, ip0x%x size0x%x",
+	DMSG("RISAF %#"PRIxPA" version %"PRIu32".%"PRIu32", ip%#"PRIx32" size%#"PRIx32,
 	     risaf->pdata.base.pa, version.major, version.minor, version.ip_id,
 	     version.size_id);
 }
 
-/*
- * @brief  Lock the RISAF IP registers for a given instance.
- * @param  risaf: RISAF instance.
- * @retval 0 if OK, negative value else.
- */
 static __maybe_unused
-int stm32_risaf_lock(struct stm32_risaf_instance *risaf)
+void stm32_risaf_lock(struct stm32_risaf_instance *risaf)
 {
 	assert(risaf);
 
 	io_setbits32(risaf_base(risaf) + _RISAF_CR, _RISAF_CR_GLOCK);
-
-	return 0;
 }
 
-/*
- * @brief  Get the RISAF lock state for a given instance.
- * @param  risaf: RISAF instance.
- *         state: lock state, true if locked, false else.
- * @retval 0 if OK, negative value else.
- */
 static __maybe_unused
-int stm32_risaf_is_locked(struct stm32_risaf_instance *risaf, bool *state)
+void stm32_risaf_is_locked(struct stm32_risaf_instance *risaf, bool *state)
 {
 	assert(risaf);
 
 	*state = (io_read32(risaf_base(risaf) + _RISAF_CR) &
 		  _RISAF_CR_GLOCK) == _RISAF_CR_GLOCK;
-
-	return 0;
 }
 
 static TEE_Result stm32_risaf_init_ddata(struct stm32_risaf_instance *risaf)
 {
 	vaddr_t base = risaf_base(risaf);
-	uint32_t hwcfgr = 0;
+	uint32_t granularity = 0;
 	uint32_t mask_lsb = 0;
 	uint32_t mask_msb = 0;
-	uint32_t granularity = 0;
+	uint32_t hwcfgr = 0;
 
 	risaf->ddata = calloc(1, sizeof(*risaf->ddata));
 	if (!risaf->ddata)
@@ -493,58 +486,23 @@ static TEE_Result stm32_risaf_init_ddata(struct stm32_risaf_instance *risaf)
 	return TEE_SUCCESS;
 }
 
-TEE_Result stm32_risaf_reconfigure(paddr_t base)
-{
-	struct stm32_risaf_instance *risaf = NULL;
-
-	SLIST_FOREACH(risaf, &risaf_list, link) {
-		struct stm32_risaf_region *regions = risaf->pdata.regions;
-		TEE_Result res = TEE_ERROR_GENERIC;
-		unsigned int i = 0;
-
-		if (base != risaf->pdata.base.pa)
-			continue;
-
-		res = clk_enable(risaf->pdata.clock);
-		if (res)
-			return res;
-
-		for (i = 0; i < risaf->pdata.nregions; i++) {
-			uint32_t id = _RISAF_GET_REGION_ID(regions[i].cfg);
-			uint32_t cfg = _RISAF_GET_REGION_CFG(regions[i].cfg);
-			uint32_t cid_cfg =
-				_RISAF_GET_REGION_CID_CFG(regions[i].cfg);
-			uintptr_t start_addr = regions[i].addr;
-			uintptr_t end_addr = start_addr + regions[i].len - 1U;
-
-			if (risaf_configure_region(risaf, id, cfg, cid_cfg,
-						   start_addr, end_addr))
-				panic();
-		}
-
-		clk_disable(risaf->pdata.clock);
-	}
-
-	return TEE_SUCCESS;
-}
-
 static TEE_Result stm32_risaf_pm_resume(struct stm32_risaf_instance *risaf)
 {
 	struct stm32_risaf_region *regions = risaf->pdata.regions;
 	size_t i = 0;
 
 	for (i = 0; i < risaf->pdata.nregions; i++) {
-		uint32_t cfg = 0;
-		uint32_t cid_cfg = 0;
-		uintptr_t start_addr = 0;
-		uintptr_t end_addr = 0;
 		uint32_t id = _RISAF_GET_REGION_ID(regions[i].cfg);
+		paddr_t start_addr = 0;
+		paddr_t end_addr = 0;
+		uint32_t cid_cfg = 0;
+		uint32_t cfg = 0;
 
 		if (!id)
 			continue;
 
-		cfg = _RISAF_GET_REGION_CFG(regions[i].cfg);
-		cid_cfg = _RISAF_GET_REGION_CID_CFG(regions[i].cfg);
+		cfg = stm32_risaf_get_region_config(regions[i].cfg);
+		cid_cfg = stm32_risaf_get_region_cid_config(regions[i].cfg);
 		start_addr = regions[i].addr;
 		end_addr = start_addr + regions[i].len - 1U;
 		if (risaf_configure_region(risaf, id, cfg, cid_cfg,
@@ -561,149 +519,95 @@ static TEE_Result stm32_risaf_pm_suspend(struct stm32_risaf_instance *risaf)
 	size_t i = 0;
 
 	for (i = 0; i < risaf->pdata.nregions; i++) {
-		uint32_t cfg = 0;
-		uint32_t risaf_en = 0;
-		uint32_t risaf_sec = 0;
-		uint32_t risaf_enc = 0;
-		uint32_t risaf_priv = 0;
+		uint32_t id = _RISAF_GET_REGION_ID(risaf->pdata.regions[i].cfg);
+		struct stm32_risaf_region *region = risaf->pdata.regions + i;
+		paddr_t start_addr = 0;
+		paddr_t end_addr = 0;
 		uint32_t cid_cfg = 0;
+		uint32_t priv = 0;
 		uint32_t rden = 0;
 		uint32_t wren = 0;
-		uintptr_t start_addr = 0;
-		uintptr_t end_addr = 0;
-		uint32_t id = _RISAF_GET_REGION_ID(risaf->pdata.regions[i].cfg);
+		uint32_t cfg = 0;
+		uint32_t enc = 0;
+		uint32_t sec = 0;
+		uint32_t en = 0;
 
 		/* Skip region not defined in DT, not configured in probe */
 		if (!id)
 			continue;
 
 		cfg = io_read32(base + _RISAF_REG_CFGR(id));
-		risaf_en = (cfg & _RISAF_REG_CFGR_BREN) << DT_RISAF_EN_SHIFT;
-		risaf_sec = ((cfg & _RISAF_REG_CFGR_SEC) >>
-			     _RISAF_REG_CFGR_SEC_SHIFT) << DT_RISAF_SEC_SHIFT;
+		en = cfg & _RISAF_REG_CFGR_BREN;
+		sec = (cfg & _RISAF_REG_CFGR_SEC) >> _RISAF_REG_CFGR_SEC_SHIFT;
+		enc = (cfg & _RISAF_REG_CFGR_ENC) >> _RISAF_REG_CFGR_ENC_SHIFT;
+		priv = (cfg & _RISAF_REG_CFGR_PRIVC_MASK) >>
+		       _RISAF_REG_CFGR_PRIVC_SHIFT;
+
+		cid_cfg = io_read32(base + _RISAF_REG_CIDCFGR(id));
+		rden = cid_cfg & _RISAF_REG_CIDCFGR_RDENC_MASK;
+		wren = (cid_cfg & _RISAF_REG_CIDCFGR_WRENC_MASK) >>
+		       _RISAF_REG_CIDCFGR_WRENC_SHIFT;
+
 #if defined(CFG_STM32MP21)
-		risaf_enc = ((cfg & _RISAF_REG_CFGR_ENC) >>
-			     _RISAF_REG_CFGR_ENC_SHIFT) << DT_RISAF_ENC_SHIFT;
+		region->cfg = id | SHIFT_U32(en, DT_RISAF_EN_SHIFT) |
+			      SHIFT_U32(sec, DT_RISAF_SEC_SHIFT) |
+			      SHIFT_U32(enc, DT_RISAF_ENC_SHIFT) |
+			      SHIFT_U32(priv, DT_RISAF_PRIV_SHIFT) |
+			      SHIFT_U32(rden, DT_RISAF_READ_SHIFT) |
+			      SHIFT_U32(wren, DT_RISAF_WRITE_SHIFT);
 #else /* defined(CFG_STM32MP21) */
-		risaf_enc = ((cfg & _RISAF_REG_CFGR_ENC) >>
-			     _RISAF_REG_CFGR_ENC_SHIFT) <<
-			    (DT_RISAF_ENC_SHIFT + 1);
+		region->cfg = id | SHIFT_U32(en, DT_RISAF_EN_SHIFT) |
+			      SHIFT_U32(sec, DT_RISAF_SEC_SHIFT) |
+			      SHIFT_U32(enc, DT_RISAF_ENC_SHIFT + 1) |
+			      SHIFT_U32(priv, DT_RISAF_PRIV_SHIFT) |
+			      SHIFT_U32(rden, DT_RISAF_READ_SHIFT) |
+			      SHIFT_U32(wren, DT_RISAF_WRITE_SHIFT);
 #endif /* defined(CFG_STM32MP21) */
-		risaf_priv = ((cfg & _RISAF_REG_CFGR_PRIVC_MASK) >>
-			      _RISAF_REG_CFGR_PRIVC_SHIFT) <<
-			     DT_RISAF_PRIV_SHIFT;
-		cid_cfg = io_read32(base + _RISAF_REG_CIDCFGR(id));
-		rden = (cid_cfg & _RISAF_REG_CIDCFGR_RDENC_MASK) <<
-		       DT_RISAF_READ_SHIFT;
-		wren = ((cid_cfg & _RISAF_REG_CIDCFGR_WRENC_MASK) >>
-			_RISAF_REG_CIDCFGR_WRENC_SHIFT) << DT_RISAF_WRITE_SHIFT;
-		risaf->pdata.regions[i].cfg = id | risaf_en | risaf_sec |
-					      risaf_enc | risaf_priv |
-					      rden | wren;
 		start_addr = io_read32(base + _RISAF_REG_STARTR(id));
 		end_addr = io_read32(base + _RISAF_REG_ENDR(id));
-		risaf->pdata.regions[i].addr = start_addr +
-					       risaf->pdata.mem_base;
-		risaf->pdata.regions[i].len = end_addr - start_addr + 1U;
+		region->addr = start_addr + risaf->pdata.mem_base;
+		region->len = end_addr - start_addr + 1;
 	}
 
 	return TEE_SUCCESS;
 }
 
 static TEE_Result
-stm32_risaf_pm(enum pm_op op, unsigned int pm_hint __unused,
+stm32_risaf_pm(enum pm_op op, unsigned int pm_hint,
 	       const struct pm_callback_handle *pm_handle)
 {
 	struct stm32_risaf_instance *risaf = pm_handle->handle;
 	TEE_Result res = TEE_ERROR_GENERIC;
 
+	assert(risaf);
+
 	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
 		return TEE_SUCCESS;
 
-	res = clk_enable(risaf->pdata.clock);
-	if (res)
-		return res;
-
 	if (op == PM_OP_RESUME)
 		res = stm32_risaf_pm_resume(risaf);
 	else
 		res = stm32_risaf_pm_suspend(risaf);
 
-	clk_disable(risaf->pdata.clock);
-
 	return res;
 }
 
 static TEE_Result stm32_risaf_acquire_access(struct firewall_query *fw,
-					     paddr_t paddr __unused,
-					     size_t size __unused, bool read,
-					     bool write)
+					     paddr_t paddr, size_t size,
+					     bool read, bool write)
 {
 	struct stm32_risaf_instance *risaf = NULL;
-	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
+	struct stm32_risaf_region *region = NULL;
 	uint32_t cidcfgr = 0;
+	unsigned int i = 0;
 	uint32_t cfgr = 0;
 	vaddr_t base = 0;
 	uint32_t id = 0;
 
-	assert(fw->ctrl->priv);
-
-	risaf = fw->ctrl->priv;
-	base = risaf_base(risaf);
-
-	if (fw->arg_count != 1)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	/*
-	 * RISAF region configuration, we assume the query is as
-	 * follows:
-	 * firewall->args[0]: Region configuration
-	 */
-	id = _RISAF_GET_REGION_ID(fw->args[0]);
-
-	if (!id || id >= risaf->pdata.nregions)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	res = clk_enable(risaf->pdata.clock);
-	if (res)
-		return res;
-
-	cfgr = io_read32(base + _RISAF_REG_CFGR(id));
-	cidcfgr = io_read32(base + _RISAF_REG_CIDCFGR(id));
-
-	/*
-	 * Access is denied if the region is disabled and OP-TEE does not run as
-	 * TDCID, or the region is not secure, or if it is not accessible in
-	 * read and/or write mode, if requested, by OP-TEE CID.
-	 */
-	if ((!(cfgr & _RISAF_REG_CFGR_BREN) && !is_tdcid) ||
-	    !(cfgr & _RISAF_REG_CFGR_SEC) ||
-	    (cidcfgr &&
-	     ((read && !_RISAF_REG_READ_OK(cidcfgr, RIF_CID1)) ||
-	      (write && !_RISAF_REG_WRITE_OK(cidcfgr, RIF_CID1)))))
-		return TEE_ERROR_ACCESS_DENIED;
-
-	clk_disable(risaf->pdata.clock);
+	assert(fw->ctrl->priv && (read || write));
 
-	return TEE_SUCCESS;
-}
-
-static TEE_Result stm32_risaf_check_access(struct firewall_query *fw,
-					   paddr_t paddr __unused,
-					   size_t size __unused,
-					   bool read __unused,
-					   bool write __unused)
-{
-	struct stm32_risaf_instance *risaf = NULL;
-	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
-	uint32_t q_config = 0;
-	uint32_t cid_list = 0;
-	uint32_t cidcfgr = 0;
-	uint32_t cfgr = 0;
-	vaddr_t base = 0;
-	uint32_t id = 0;
-
-	assert(fw->ctrl->priv);
+	if (paddr == TZDRAM_BASE && size == TZDRAM_SIZE)
+		return TEE_SUCCESS;
 
 	risaf = fw->ctrl->priv;
 	base = risaf_base(risaf);
@@ -716,15 +620,21 @@ static TEE_Result stm32_risaf_check_access(struct firewall_query *fw,
 	 * follows:
 	 * firewall->args[0]: Region configuration
 	 */
-	q_config = fw->args[0];
-	id = _RISAF_GET_REGION_ID(q_config);
+	id = _RISAF_GET_REGION_ID(fw->args[0]);
 
 	if (!id || id >= risaf->pdata.nregions)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	res = clk_enable(risaf->pdata.clock);
-	if (res)
-		return res;
+	for (i = 0; i < risaf->pdata.nregions; i++) {
+		if (id == _RISAF_GET_REGION_ID(risaf->pdata.regions[i].cfg)) {
+			region = &risaf->pdata.regions[i];
+			if (region->addr != paddr || region->len != size)
+				return TEE_ERROR_GENERIC;
+			break;
+		}
+	}
+	if (!region)
+		return TEE_ERROR_ITEM_NOT_FOUND;
 
 	cfgr = io_read32(base + _RISAF_REG_CFGR(id));
 	cidcfgr = io_read32(base + _RISAF_REG_CIDCFGR(id));
@@ -734,27 +644,15 @@ static TEE_Result stm32_risaf_check_access(struct firewall_query *fw,
 	 * TDCID, or the region is not secure, or if it is not accessible in
 	 * read and/or write mode, if requested, by OP-TEE CID.
 	 */
-	if ((!(cfgr & _RISAF_REG_CFGR_BREN) && !is_tdcid) ||
-	    ((_RISAF_GET_REGION_CFG(q_config) & _RISAF_REG_CFGR_SEC) &&
-	     !(cfgr & _RISAF_REG_CFGR_SEC)))
-		return TEE_ERROR_ACCESS_DENIED;
-
-	cid_list = _RISAF_GET_REGION_CFG(q_config) &
-		   _RISAF_REG_CFGR_PRIVC_MASK;
-	if (!(cid_list == (cfgr & cid_list)))
+	if (!(cfgr & _RISAF_REG_CFGR_BREN) && !is_tdcid)
 		return TEE_ERROR_ACCESS_DENIED;
 
-	cid_list = _RISAF_GET_REGION_CID_CFG(q_config) &
-		   _RISAF_REG_CIDCFGR_RDENC_MASK;
-	if (cid_list != (cidcfgr & cid_list))
+	if ((cfgr & _RISAF_REG_CFGR_BREN) &&
+	    (!(cfgr & _RISAF_REG_CFGR_SEC) ||
+	     (read && !_RISAF_REG_READ_OK(cidcfgr, RIF_CID1)) ||
+	     (write && !_RISAF_REG_WRITE_OK(cidcfgr, RIF_CID1)))) {
 		return TEE_ERROR_ACCESS_DENIED;
-
-	cid_list = _RISAF_GET_REGION_CID_CFG(q_config) &
-		   _RISAF_REG_CIDCFGR_WRENC_MASK;
-	if (cid_list != (cidcfgr & cid_list))
-		return TEE_ERROR_ACCESS_DENIED;
-
-	clk_disable(risaf->pdata.clock);
+	}
 
 	return TEE_SUCCESS;
 }
@@ -766,7 +664,7 @@ static TEE_Result stm32_risaf_reconfigure_region(struct firewall_query *fw,
 	struct stm32_risaf_region *region = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	uint32_t exceptions = 0;
-	uint32_t q_config = 0;
+	uint32_t q_cfg = 0;
 	unsigned int i = 0;
 	uint32_t id = 0;
 
@@ -782,8 +680,8 @@ static TEE_Result stm32_risaf_reconfigure_region(struct firewall_query *fw,
 	 * follows:
 	 * firewall->args[0]: Region configuration
 	 */
-	q_config = fw->args[0];
-	id = _RISAF_GET_REGION_ID(q_config);
+	q_cfg = fw->args[0];
+	id = _RISAF_GET_REGION_ID(q_cfg);
 
 	if (!id || id >= risaf->pdata.nregions)
 		return TEE_ERROR_BAD_PARAMETERS;
@@ -799,47 +697,40 @@ static TEE_Result stm32_risaf_reconfigure_region(struct firewall_query *fw,
 	if (!region)
 		return TEE_ERROR_ITEM_NOT_FOUND;
 
-	res = clk_enable(risaf->pdata.clock);
-	if (res)
-		return res;
-
 	DMSG("Reconfiguring %s region ID: %"PRIu32, risaf->pdata.risaf_name,
 	     id);
 
 	exceptions = cpu_spin_lock_xsave(&risaf->pdata.conf_lock);
 	res = risaf_configure_region(risaf, id,
-				     _RISAF_GET_REGION_CFG(q_config),
-				     _RISAF_GET_REGION_CID_CFG(q_config),
+				     stm32_risaf_get_region_config(q_cfg),
+				     stm32_risaf_get_region_cid_config(q_cfg),
 				     region->addr,
 				     region->addr + region->len - 1);
 
 	cpu_spin_unlock_xrestore(&risaf->pdata.conf_lock, exceptions);
 
-	clk_disable(risaf->pdata.clock);
-
 	return res;
 }
 
 static const struct firewall_controller_ops firewall_ops = {
 	.acquire_memory_access = stm32_risaf_acquire_access,
-	.check_memory_access = stm32_risaf_check_access,
 	.set_memory_conf = stm32_risaf_reconfigure_region,
 };
 
 static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 				    const void *compat_data)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
-	struct dt_node_info dt_info = { };
+	const struct stm32_risaf_compat_data *compat = compat_data;
+	struct firewall_controller *controller = NULL;
 	struct stm32_risaf_instance *risaf = NULL;
 	struct stm32_risaf_region *regions = NULL;
-	struct firewall_controller *controller = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info dt_info = { };
+	const fdt32_t *conf_list = NULL;
+	const fdt64_t *cuint = NULL;
 	unsigned int nregions = 0;
-	int len = 0;
 	unsigned int i = 0;
-	const fdt64_t *cuint = NULL;
-	const fdt32_t *conf_list = NULL;
-	const struct stm32_risaf_compat_data *compat = compat_data;
+	int len = 0;
 
 	res = stm32_rifsc_check_tdcid(&is_tdcid);
 	if (res)
@@ -870,12 +761,15 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 
 	conf_list = fdt_getprop(fdt, node, "memory-region", &len);
 	if (!conf_list) {
-		DMSG("RISAF %#"PRIxPTR": No configuration in DT, use default",
+		DMSG("RISAF %#"PRIxPA": No configuration in DT, use default",
 		     risaf->pdata.base.pa);
 		free(risaf);
 		return TEE_SUCCESS;
 	}
 
+	nregions = (unsigned int)len / sizeof(uint32_t);
+
+	/* Silently allow unexpected truncated names */
 	strncpy(risaf->pdata.risaf_name, fdt_get_name(fdt, node, NULL),
 		sizeof(risaf->pdata.risaf_name) - 1);
 
@@ -889,44 +783,58 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 
 	risaf_print_version(risaf);
 
-	cuint = fdt_getprop(fdt, node, "st,mem-map", NULL);
-	if (!cuint)
+	cuint = fdt_getprop(fdt, node, "st,mem-map", &len);
+	if (!cuint || (size_t)len != sizeof(*cuint) * 2)
 		panic();
 
-	risaf->pdata.mem_base = (uintptr_t)fdt64_to_cpu(*cuint);
+	risaf->pdata.mem_base = (paddr_t)fdt64_to_cpu(*cuint);
 	risaf->pdata.mem_size = (size_t)fdt64_to_cpu(*(cuint + 1));
 
-	nregions = (unsigned int)len / sizeof(uint32_t);
-
 	regions = calloc(nregions, sizeof(*regions));
 	if (nregions && !regions) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto err_ddata;
 	}
 
+	DMSG("RISAF %#"PRIxPA" memory range: %#"PRIxPA" - %#"PRIxPA,
+	     risaf->pdata.base.pa, risaf->pdata.mem_base,
+	     risaf->pdata.mem_base + risaf->pdata.mem_size - 1);
+
 	for (i = 0; i < nregions; i++) {
-		int pnode = 0;
 		const fdt32_t *prop = NULL;
-		uint32_t id = 0;
-		uint32_t cfg = 0;
+		paddr_t start_addr = 0;
+		paddr_t end_addr = 0;
 		uint32_t cid_cfg = 0;
-		uintptr_t start_addr = 0;
-		uintptr_t end_addr = 0;
+		uint32_t phandle = 0;
+		uint32_t cfg = 0;
+		uint32_t id = 0;
+		int pnode = 0;
 
-		pnode =
-		fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(*(conf_list + i)));
+		phandle = fdt32_to_cpu(*(conf_list + i));
+		pnode = fdt_node_offset_by_phandle(fdt, phandle);
 		if (pnode < 0)
 			continue;
 
-		regions[i].addr = (uintptr_t)fdt_reg_base_address(fdt, pnode);
+		regions[i].addr = fdt_reg_base_address(fdt, pnode);
 		regions[i].len = fdt_reg_size(fdt, pnode);
+		if (regions[i].addr == DT_INFO_INVALID_REG ||
+		    regions[i].len == DT_INFO_INVALID_REG_SIZE) {
+			EMSG("Invalid config in node %s",
+			     fdt_get_name(fdt, pnode, NULL));
+			panic();
+		}
 
 		if (!regions[i].len)
 			continue;
 
-		/* Skip 'op_tee' region, in which OP-TEE code is executed */
+		/*
+		 * The secure bootloader is in charge of configuring RISAF
+		 * related to OP-TEE secure memory. Therefore, skip OP-TEE
+		 * region so that RISAF configuration cannot interfere with
+		 * OP-TEE execution flow.
+		 */
 		if (regions[i].addr == TZDRAM_BASE &&
-		    regions[i].len == TZDRAM_RESERVED_SIZE) {
+		    regions[i].len == TZDRAM_SIZE) {
 			continue;
 		}
 
@@ -936,10 +844,6 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 
 		regions[i].cfg = fdt32_to_cpu(*prop);
 
-		DMSG("RISAF %#"PRIxPTR": cfg 0x%08x - addr 0x%08lx - len 0x%08lx",
-		     risaf->pdata.base.pa, regions[i].cfg, regions[i].addr,
-		     regions[i].len);
-
 		if (risaf_check_region_boundaries(risaf, &regions[i]) ||
 		    risaf_check_overlap(risaf, regions, i))
 			panic();
@@ -947,9 +851,9 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 		id = _RISAF_GET_REGION_ID(regions[i].cfg);
 		assert(id < risaf->ddata->max_base_regions);
 
-		cfg = _RISAF_GET_REGION_CFG(regions[i].cfg);
+		cfg = stm32_risaf_get_region_config(regions[i].cfg);
 
-		cid_cfg = _RISAF_GET_REGION_CID_CFG(regions[i].cfg);
+		cid_cfg = stm32_risaf_get_region_cid_config(regions[i].cfg);
 
 		start_addr = regions[i].addr;
 		end_addr = start_addr + regions[i].len - 1U;
@@ -959,8 +863,6 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 			panic();
 	}
 
-	clk_disable(risaf->pdata.clock);
-
 	controller = calloc(1, sizeof(*controller));
 	if (!controller)
 		panic();
@@ -995,11 +897,11 @@ err:
 static const struct dt_device_match risaf_match_table[] = {
 	{
 		.compatible = "st,stm32mp25-risaf",
-		.compat_data = (void *)&stm32_risaf_compat,
+		.compat_data = &stm32_risaf_compat,
 	},
 	{
 		.compatible = "st,stm32mp25-risaf-enc",
-		.compat_data = (void *)&stm32_risaf_enc_compat,
+		.compat_data = &stm32_risaf_enc_compat,
 	},
 	{ }
 };
diff --git a/core/include/drivers/stm32_risaf.h b/core/include/drivers/stm32_risaf.h
index 0970b95b1..1d611f013 100644
--- a/core/include/drivers/stm32_risaf.h
+++ b/core/include/drivers/stm32_risaf.h
@@ -1,24 +1,23 @@
-/* SPDX-License-Identifier: BSD-3-Clause */
+/* SPDX-License-Identifier: BSD-2-Clause */
 /*
- * Copyright (c) 2022, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2021-2024, STMicroelectronics
  */
 
 #ifndef __DRIVERS_STM32_RISAF_H__
 #define __DRIVERS_STM32_RISAF_H__
 
-#include <tee_api_types.h>
-#include <types_ext.h>
-
-TEE_Result stm32_risaf_reconfigure(paddr_t base);
-
+/*
+ * stm32_risaf_clear_illegal_access_flags() - Clears flags raised when an
+ * illegal access occurs on a memory region
+ */
 void stm32_risaf_clear_illegal_access_flags(void);
 
-#ifdef CFG_TEE_CORE_DEBUG
-void stm32_risaf_dump_erroneous_data(void);
-#else /* CFG_TEE_CORE_DEBUG */
-static inline void stm32_risaf_dump_erroneous_data(void)
-{
-}
-#endif /* CFG_TEE_CORE_DEBUG */
+/*
+ * stm32_risaf_print_erroneous_data() - Prints the data associated to an illegal
+ * access occurring on a memory protected by a RISAF : faulty address and
+ * firewall attributes of the master causing the illegal access. This function
+ * is stubbed when CFG_TEE_CORE_DEBUG is disabled.
+ */
+void stm32_risaf_print_erroneous_data(void);
 
 #endif /*__DRIVERS_STM32_RISAF_H__*/
-- 
2.47.2

